File: backend/internal/domain/position.go
package domain

type Position struct {
	X int
	Y int
}

func (p Position) IsValid() bool {
	return p.X >= 0 && p.X < BoardSize && p.Y >= 0 && p.Y < BoardSize
}

func (p Position) Offset(dx, dy int) Position {
	return Position{X: p.X + dx, Y: p.Y + dy}
}

File: backend/src/Caro.Domain/Position.cs
namespace Caro.Domain;

public readonly record struct Position(int X, int Y)
{
    public bool IsValid() =>
        X >= 0 && X < Constants.Board.Size && Y >= 0 && Y < Constants.Board.Size;

    public Position Offset(int dx, int dy) => new(X + dx, Y + dy);
}

public readonly record struct Cell(int X, int Y, Player Player)
{
    public bool IsEmpty() => Player == Player.None;
}
